Add support for direct email of output

kpfleming opened the pull request:
Adds an 'email' section to configuration, allowing the tool to directly
send the filtered/formatted output via 'smtplib', avoiding the need to
run the tool in a cron-type wrapper to capture the output.

kpfleming commented:
This does not have additional tests yet; I would appreciate some guidance as to how tests should be integrated into the existing structure.

kpfleming commented:
I'm running this in `ExecStopPost` actions in systemd service units, and I don't want to have to run that through a shell (to provide piping) if I can avoid it. The suppress_empty logic would be another reason.

There aren't any new module depdencies, as both `smtplib` and `ssl` are part of the Python standard library.

I would be OK calling out to the 'mail' interface as an option; that would allow some users to use local (non-SMTP) mail delivery, or any other delivery they want, but still allow other users to use direct SMTP delivery (which I prefer). This could be selected based on the presence of the `command` key in the `email` map.

kpfleming commented:
Oh... another reason: I don't want the cursor to be updated if the attempt to send the email fails. I already ran into that problem during development and couldn't figure out what was going on :-)

twaugh commented:
Ah, hadn't realised they are part of the standard library.

Something I found while playing with this locally: it can be useful to be able to run the command and see the output in the terminal. Maybe there should be a command line option for enabling/disabling emailing the output? Commonly `-c` is for this type of thing (e.g. gzip).

twaugh commented:
Should `--dry-run` still send email, do you think? Currently it does.

kpfleming commented:
I think you answered your first question with your second question :-)

I'll add support for command-based email delivery, and also ensure that --dry-run displays the constructed email and information about how it would be delivered if email mode is configured (but doesn't send it).

kpfleming commented:
New version pushed; adds 'command' email output type, refactored email-sending code, and changed the structure of the email configuration map. Also directly prints email that would be sent during --dry-run instead of sending it.

kpfleming commented:
This now includes substantial testing of the email support, both configuration parsing and email generation.

twaugh commented on journal_brief/cli/main.py:
```diff
@@ -17,17 +17,25 @@
                            list_formatters,
                            JournalFilter)
+from journal_brief.cli.constants import EMAIL_SUPPRESS_EMPTY_TEXT
+from journal_brief.cli.constants import EMAIL_DRY_RUN_SEPARATOR
```
Minor nit: could use `from ... import (..., ...)` here (and other places we import these).

kpfleming replied:
Done.

twaugh commented on journal_brief/cli/main.py:
```diff
@@ -200,6 +208,45 @@ def get_formatters(self):
+    def send_email(self, output):
+        email = self.config.get('email')
+
+        if len(output) == 0:
```
Minor nit: could use `if not output:` here.

kpfleming replied:
Done.

twaugh commented:
Nice, thanks!

kpfleming commented:
Next up: an article for opensource.com showing how to use this feature :-)

twaugh commented:
\o/
